src/main/java/com/uid2/shared/attest/AttestationTokenRetriever.java

```diff
@@ -0,0 +1,205 @@
+import io.vertx.core.json.JsonObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.crypto.Cipher;
+import java.io.IOException;
+import java.net.*;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.nio.charset.StandardCharsets;
+import java.security.*;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.Map;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class AttestationTokenRetriever {
+    private static final Logger LOGGER = LoggerFactory.getLogger(UidCoreClient.class);
+    private final IAttestationProvider attestationProvider;
+    private final ApplicationVersion appVersion;
+    private AtomicReference<String> attestationToken;
+    private Handler<Integer> responseWatcher;
+    private final String attestationEndpoint;
+    private final HttpClient httpClient;
+    private final IClock clock;
+    private ScheduledThreadPoolExecutor executor;
+    // Set this to be Instant.MAX so that if it's not set it won't trigger the re-attest
+    private Instant attestationTokenExpiresAt = Instant.MAX;
+    private final Lock lock;
+    private final AttestationTokenDecryptor attestationTokenDecryptor;
+
+
+    public AttestationTokenRetriever(String attestationEndpoint, ApplicationVersion appVersion, IAttestationProvider attestationProvider,
+                                     Handler<Integer> responseWatcher, IClock clock, HttpClient httpClient,
+                                     AttestationTokenDecryptor attestationTokenDecryptor, ScheduledThreadPoolExecutor executor) throws IOException {
+        this.attestationEndpoint = attestationEndpoint;
+        this.appVersion = appVersion;
+        this.attestationProvider = attestationProvider;
+        this.attestationToken = new AtomicReference<>(null);
+        this.responseWatcher = responseWatcher;
+        this.clock = clock;
+        this.lock = new ReentrantLock();
+        if (httpClient == null) {
+            this.httpClient = HttpClient.newHttpClient();
+        } else {
+            this.httpClient = httpClient;
+        }
+        if (attestationTokenDecryptor == null) {
+            this.attestationTokenDecryptor = new AttestationTokenDecryptor();
+        } else {
+            this.attestationTokenDecryptor = attestationTokenDecryptor;
+        }
+        if (executor == null) {
+            this.executor = new ScheduledThreadPoolExecutor(1);
+        } else {
+            this.executor = executor;
+        }
+    }
+
+    private void attestationExpirationCheck(){
+        Instant currentTime = clock.now();
+        Instant tenMinutesBeforeExpire = attestationTokenExpiresAt.minusSeconds(600);
+
+        stopAttestationExpirationCheck();
```

Comment:
If attestation fails for any reason, this will mean it has stopped and it will not reattest. I think this should turn off the scheduler, do the reattest attempt, and then turn on the scheduler again, even if attestation has failed.